internal/gps/prune.go

```diff
@@ -107,152 +111,141 @@ func pruneNestedVendorDirs(baseDir string) error {
 	return filepath.Walk(baseDir, stripVendor)
 }
 
-// pruneUnusedPackages deletes unimported packages found within baseDir.
+// pruneVendorDirs deletes all nested vendor directories within baseDir.
+// func pruneVendorDirs(fs filesystemState, logger *log.Logger) error {
+// 	toDelete := collectNestedVendorDirs(fs)
+
+// 	for _, path := range toDelete {
+// 		logger.Printf("  * %s", path)
+
+// 		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
+// 			return err
+// 		}
+// 	}
+
+// 	return nil
+// }
+
+// func collectNestedVendorDirs(fs filesystemState) []string {
+// 	toDelete := make([]string, 0, len(fs.dirs)/4)
+
+// 	for _, dir := range fs.dirs {
+// 		if filepath.Base(dir)
+// 		toDelete = append(toDelete, dir)
+// 	}
+
+// 	for _, link := range fs.links {
+// 		toDelete = append(toDelete)
+// 	}
+
+// 	return files
+// }
+
+// pruneUnusedPackages deletes unimported packages found in fs.
 // Determining whether packages are imported or not is based on the passed LockedProject.
-func pruneUnusedPackages(lp LockedProject, projectDir string, logger *log.Logger) error {
+func pruneUnusedPackages(lp LockedProject, fs filesystemState, logger *log.Logger) error {
 	pr := string(lp.Ident().ProjectRoot)
-	logger.Printf("Calculating unused packages in %s to prune.\n", pr)
+	unusedPackages := calculateUnusedPackages(lp, fs)
 
-	unusedPackages, err := calculateUnusedPackages(lp, projectDir)
-	if err != nil {
-		return errors.Wrapf(err, "could not calculate unused packages in %s", pr)
-	}
+	logger.Printf("Found %d unused packages in %s:\n", len(unusedPackages), pr)
 
-	logger.Printf("Found the following unused packages in %s:\n", pr)
 	for pkg := range unusedPackages {
 		logger.Printf("  * %s\n", filepath.Join(pr, pkg))
 	}
 
-	unusedPackagesFiles, err := collectUnusedPackagesFiles(projectDir, unusedPackages)
-	if err != nil {
-		return errors.Wrapf(err, "could not collect unused packages' files in %s", pr)
-	}
+	toDelete := collectUnusedPackagesFiles(fs, unusedPackages)
+
+	logger.Printf("Deleting %d files in unused packages:", len(toDelete))
 
-	if err := deleteFiles(unusedPackagesFiles); err != nil {
-		return errors.Wrapf(err, "")
+	for _, path := range toDelete {
+		logger.Printf("  * %s", path)
+
+		if err := os.Remove(path); err != nil && !os.IsNotExist(err) {
```

Comment:
how are we ordering the deletion of directories such that we don't leave unnecessary empty ones around, but still do it in the correct order so that directories are only removed once all of their contents have been removed?

Reply:
This is one of the most annoying parts of pruning for me.

I opted to only delete files in the new prune functions, and empty directories are kept behind.
I think empty directories as a minor issue atm since `git` and `hg` ignore empty directories by default (not sure how `bzr` handles them).

We'll need to replace `filesystemState` with a hierarchical representation to allow determining which directories are empty efficiently or come up with a different solution.

Reply:
> We'll need to replace filesystemState with a hierarchical representation to allow determining which directories are empty efficiently or come up with a different solution.

If the paths are normalized, can the list just be sorted so that everything is deleted bottom-up? Then whenever the next item is a new dir, empty-check the previous.

Reply:
Dirs are not added to the list currently. The logic to determine when to add a dir to the list will be more involved with the current struct.

Reply:
They don't need to be in the list for what I was describing.  I just meant that if the next file is in a different dir then the previous file, then it was the last file to delete from that dir (assuming a proper normalized sort), so empty-check it.  This doesn't account for directories which were already empty before any pruning, but that seems like a weird edge case.

Reply:
> I think empty directories as a minor issue atm

we _may_ be able to skate by in an initial implementation without them because of git/hg's treatment of empty dirs, but i suspect we'll need to catch up rather quickly. beyond the general annoyance empty dirs will cause for folks, i suspect that correct implementation of #1113 will also demand that we do something a little more sophisticated here, anyway.

i don't have a clear implementation in mind yet, but i have some vague ideas swirling around regarding prefix-based iteration/exploration of a trie.

Reply:
I've 2 approaches in mind to do this:

1. Loop through all dirs in the `filesystemState` struct and check if each dir is empty or not. This can be done once after all options are done, or after each option. I prefer doing it once at the end.
2. After each option deletes its files, determine the touched dirs from the slice containing the `toDelete` files, and loop through only those dirs and check for emptiness.

The key operation that I'm trying to minimize is opening the file to use `os.File.Readdirname`.
I can keep looping over the slices to find files in a dir, but a prefix tree would make that easier. Giving `filesystemState` hierarchy could be even better and more specialized.

But I'm trying to resist the temptation to change `filesystemState` to a more complex struct that provides more functionality by finding a simpler solution.

Reply:
just off the cuff, but maybe it's as simple as a sort function, where records represent dirs and their contents, and dirs that are subpaths of other dir records are moved to the front? that way you can proceed linearly through the sorted slice, assured that you will never visit a dir before any of its ancestors.